src/peer-store/key-book.js

```diff
@@ -0,0 +1,83 @@
+'use strict'
+
+const errcode = require('err-code')
+const debug = require('debug')
+const log = debug('libp2p:peer-store:key-book')
+log.error = debug('libp2p:peer-store:key-book:error')
+
+const PeerId = require('peer-id')
+
+const Book = require('./book')
+
+const {
+  codes: { ERR_INVALID_PARAMETERS }
+} = require('../errors')
+
+/**
+ * The KeyBook is responsible for keeping the known public keys of a peer.
+ */
+class KeyBook extends Book {
+  /**
+  * @constructor
+  * @param {PeerStore} peerStore
+  */
+  constructor (peerStore) {
+    super({
+      peerStore,
+      eventName: 'change:pubkey', // TODO: the name is not probably the best!?
+      eventProperty: 'pubkey',
+      eventTransformer: (data) => data.pubKey
+    })
+
+    /**
+     * Map known peers to their known Public Key.
+     * @type {Map<string, PeerId>}
+     */
+    this.data = new Map()
+  }
+
+  /**
+   * Set PeerId. If the peer was not known before, it will be added.
+   * @override
+   * @param {PeerId} peerId
+   * @return {KeyBook}
+ */
+  set (peerId) {
+    if (!PeerId.isPeerId(peerId)) {
+      log.error('peerId must be an instance of peer-id to store data')
+      throw errcode(new Error('peerId must be an instance of peer-id'), ERR_INVALID_PARAMETERS)
+    }
+
+    const id = peerId.toB58String()
+    const recPeerId = this.data.get(id)
+
+    !recPeerId && this._ps.emit('peer', peerId)
```

Comment:
> We should avoid other things triggering events on the peer store and instead rely on the peer store to determine when it should be emitting things. If we don't isolate we'll run the risk of having what we used to have, where several things are emitting peer events which can result in a lot of unnecessary events.

I did not understand your point here. Currently, the `Book.js` base class is emitting the events for changed peers, while this is the previous event emitted when the peer was discovered (in the addressBook), but now that we have the keyBook keeping the peerIds, it feels that this is the place this should be.

> It would probably be good to take a step back and think about the flow of the peers data into the peer store. I dont think that updating the keybook anytime we learn about the peer is the intuitive action. Updating the keybook after crypto negotiation makes sense.

> It might be good to revisit the peer store readme and make sure that is up to date and makes sense with these flows.

I added it to update the keyBook because I did not want to keep track of the peerIds in two different data structures. If we move into only storing the `peerId` in the keybook if we have a public key for it, we will need to go back with the `peer` event into the addressBook.

Updating the keybook after crypto negotiation makes sense, and I forgot that flow. However, subsystems like the dht will also need to be able to set the public key for a peer.

All things considered, should we go into only storing the PeerId when it has a public key (we can reconstruct it into the getters with `createFromCID` if we do not store it? With that, we do not update the keybook anymore on the other books set and we emit the `peer` event from the addressBook? This way, the upgrader and the dht would call the `keyBook.set()`...

Reply:
If we go this way, should we throw an error if there is no public key provided in the peerId to set? Probably we should change the API into `keyBook.set(peerId, publicKey)`. While the parameter is redundant, it will sign in an easier way to the user that it is expected